{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at the Dataset\n",
    "The purpose of this notebook is to look closer at the dataset of genes, natural language descriptions, and ontology term annotations that are used in this work. As included in the preprocessing notebooks, these data are drawn from files from either publications supplements like Oellrich, Walls et al. (2015) or model species databases such as TAIR, MaizeGDB, and SGN. The datasets are already loaded and merged using classes available through the oats package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warming up PyWSD (takes ~10 secs)... took 3.8808023929595947 secs.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import datetime\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "import math\n",
    "import sys\n",
    "import gensim\n",
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from nltk.corpus import brown\n",
    "from nltk.tokenize import word_tokenize, sent_tokenize\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.probability import FreqDist\n",
    "from gensim.parsing.preprocessing import strip_non_alphanum, stem_text, preprocess_string, remove_stopwords\n",
    "from gensim.utils import simple_preprocess\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "sys.path.append(\"../../oats\")\n",
    "sys.path.append(\"../../oats\")\n",
    "from oats.utils.utils import save_to_pickle, load_from_pickle, flatten, to_hms\n",
    "from oats.utils.utils import function_wrapper_with_duration, remove_duplicates_retain_order\n",
    "from oats.biology.dataset import Dataset\n",
    "from oats.biology.groupings import Groupings\n",
    "from oats.biology.relationships import ProteinInteractions, AnyInteractions\n",
    "from oats.annotation.ontology import Ontology\n",
    "from oats.annotation.annotation import annotate_using_noble_coder, term_enrichment\n",
    "from oats.distances import pairwise as pw\n",
    "from oats.nlp.vocabulary import get_overrepresented_tokens, get_vocab_from_tokens\n",
    "from oats.nlp.vocabulary import reduce_vocab_connected_components, reduce_vocab_linares_pontes, token_enrichment\n",
    "\n",
    "warnings.simplefilter('ignore')\n",
    "pd.set_option('display.max_rows', 500)\n",
    "pd.set_option('display.max_columns', 500)\n",
    "pd.set_option('display.width', 1000)\n",
    "nltk.download('punkt', quiet=True)\n",
    "nltk.download('brown', quiet=True)\n",
    "nltk.download('averaged_perceptron_tagger', quiet=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Paths to the files that are used for this notebook.\n",
    "plant_dataset_path = \"../../plant-data/genes_texts_annots.csv\"\n",
    "\n",
    "# Paths to files with mappings to groups.\n",
    "kegg_pathways_path = \"../../plant-data/reshaped_data/kegg_pathways.csv\" \n",
    "plantcyc_pathways_path = \"../../plant-data/reshaped_data/plantcyc_pathways.csv\" \n",
    "lloyd_meinke_subsets_path = \"../../plant-data/reshaped_data/lloyd_meinke_subsets.csv\" \n",
    "lloyd_meinke_classes_path = \"../../plant-data/reshaped_data/lloyd_meinke_classes.csv\" \n",
    "\n",
    "# Paths to files that map group identifers to longer group names.\n",
    "kegg_pathways_names_path = \"../../plant-data/reshaped_data/kegg_pathways_name_map.csv\"\n",
    "plantcyc_pathways_names_path = \"../../plant-data/reshaped_data/plantcyc_pathways_name_map.csv\"\n",
    "lloyd_meinke_subsets_names_path = \"../../plant-data/reshaped_data/lloyd_meinke_subsets_name_map.csv\"\n",
    "lloyd_meinke_classes_names_path = \"../../plant-data/reshaped_data/lloyd_meinke_classes_name_map.csv\"\n",
    "\n",
    "# Path to file with plant ortholog mappings.\n",
    "ortholog_file_path = \"../../plant-data/databases/panther/PlantGenomeOrthologs_IRB_Modified.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create and name an output directory according to when the notebooks was run.\n",
    "OUTPUT_NAME = \"composition\"\n",
    "OUTPUT_DIR = os.path.join(\"../outputs\",\"{}_{}_{}\".format(OUTPUT_NAME,datetime.datetime.now().strftime('%m_%d_%Y_h%Hm%Ms%S'),random.randrange(1000,9999)))\n",
    "os.mkdir(OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6274</td>\n",
       "      <td>3818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7907</td>\n",
       "      <td>4841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions\n",
       "0     ath                     6274                 3818\n",
       "1     gmx                       30                   23\n",
       "2     mtr                       37                   36\n",
       "3     osa                       92                   85\n",
       "4     sly                       69                   69\n",
       "5     zma                     1405                  810\n",
       "6   total                     7907                 4841"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reading in and describing the dataset of plant genes.\n",
    "plant_dataset = Dataset(plant_dataset_path)\n",
    "plant_dataset.filter_has_description()\n",
    "plant_dataset.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What's there for each species?\n",
    "The previously loaded dataset contains all of the genes that across six plant species that have natural language description data for phenotype(s) related to that gene. Each gene can have multiple descriptions annotated to it, which were combined or concatenated when the datasets from multiple sources were merged in creating the pickled datasets. Arabidopsis has the highest number of genes that satisfy this criteria, followed by maize, and then followed by the other four species which have a relatively low number of genes that satisfy this criteria, at least given the sources used for this work. Note that the number of unique descriptions is lower than the number of genes in call cases, because multiple genes can have the same phenotype description associated with them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6274</td>\n",
       "      <td>3818</td>\n",
       "      <td>30121</td>\n",
       "      <td>261441</td>\n",
       "      <td>7331</td>\n",
       "      <td>5305</td>\n",
       "      <td>261441</td>\n",
       "      <td>6792</td>\n",
       "      <td>5084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>7512</td>\n",
       "      <td>47139</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>47139</td>\n",
       "      <td>1722</td>\n",
       "      <td>498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7907</td>\n",
       "      <td>4841</td>\n",
       "      <td>38795</td>\n",
       "      <td>316616</td>\n",
       "      <td>8277</td>\n",
       "      <td>5984</td>\n",
       "      <td>316616</td>\n",
       "      <td>7663</td>\n",
       "      <td>5913</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species\n",
       "0     ath                     6274                 3818        30121       261441          7331          5305        261441           6792                      5084\n",
       "5     zma                     1405                  810         7512        47139          1846          1317         47139           1722                       498\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        97\n",
       "4     sly                       69                   69          359         1678           577           438          1678            552                        99\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       123\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12\n",
       "6   total                     7907                 4841        38795       316616          8277          5984        316616           7663                      5913"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = plant_dataset\n",
    "\n",
    "wnl = WordNetLemmatizer()\n",
    "lemmatize_doc = lambda d: [wnl.lemmatize(x) for x in simple_preprocess(d)]\n",
    "\n",
    "dists = defaultdict(list)\n",
    "\n",
    "sent_lists = {}\n",
    "token_lists = {}\n",
    "stems_lists = {}\n",
    "lemma_lists = {}\n",
    "\n",
    "\n",
    "# For each individual species.\n",
    "for species in data.get_species():\n",
    "    df = data.to_pandas()\n",
    "    subset = df[df[\"species\"]==species]\n",
    "    sentences = [sent_tokenize(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_stemmed = [preprocess_string(d) for d in subset[\"descriptions\"].values]\n",
    "    descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"descriptions\"].values]\n",
    "    sent_lists[species] = flatten(sentences)\n",
    "    token_lists[species] = flatten(descriptions_not_stemmed)\n",
    "    stems_lists[species] = flatten(descriptions_stemmed)    \n",
    "    lemma_lists[species] = flatten(descriptions_lemmatized)\n",
    "    \n",
    "    # What about the distributions of words per gene and sentences per gene?\n",
    "    dists[\"species\"].extend([species]*subset.shape[0])\n",
    "    dists[\"num_words\"].extend([len(word_tokenize(x)) for x in subset[\"descriptions\"].values])\n",
    "    dists[\"num_sents\"].extend([len(sent_tokenize(x)) for x in subset[\"descriptions\"].values])\n",
    "    \n",
    "# For the entire dataset including all of the species.\n",
    "df = data.to_pandas()\n",
    "subset = df\n",
    "sentences = [sent_tokenize(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_not_stemmed = [simple_preprocess(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_stemmed = [preprocess_string(d) for d in subset[\"descriptions\"].values]\n",
    "descriptions_lemmatized = [lemmatize_doc(d) for d in subset[\"descriptions\"].values]\n",
    "sent_lists[\"total\"] = flatten(sentences)\n",
    "token_lists[\"total\"] = flatten(descriptions_not_stemmed)\n",
    "stems_lists[\"total\"] = flatten(descriptions_stemmed)    \n",
    "lemma_lists[\"total\"] = flatten(descriptions_lemmatized)\n",
    "\n",
    "# What about lemmas that are uniquely used for a particular species?\n",
    "lemma_sets_unique_to_species = {}\n",
    "for species in data.get_species():\n",
    "    other_species = [s for s in data.get_species() if s != species]\n",
    "    lemmas_used_in_other_species = set(flatten([lemma_lists[s] for s in other_species]))\n",
    "    unique_lemmas = set(lemma_lists[species]).difference(lemmas_used_in_other_species)\n",
    "    lemma_sets_unique_to_species[species] = unique_lemmas\n",
    "lemma_sets_unique_to_species[\"total\"] = flatten([list(s) for s in lemma_sets_unique_to_species.values()])\n",
    "\n",
    "    \n",
    "# Create a dataframe to contain the summarizing information about this dataset, and sort it by number of genes.\n",
    "# Unique gene identifiers is just the total number of genes, this column name should be changed in the class...\n",
    "df = data.describe() \n",
    "condition = (df.species==\"total\")\n",
    "excluded = df[condition]\n",
    "included = df[~condition]\n",
    "df_sorted = included.sort_values(by=\"unique_gene_identifiers\", ascending=False)\n",
    "df = pd.concat([df_sorted,excluded])\n",
    "\n",
    "# Add columns summarizing information about the text descriptions in the dataset.\n",
    "df[\"total_sents\"] = df[\"species\"].map(lambda x: len(sent_lists[x]))\n",
    "df[\"total_words\"] = df[\"species\"].map(lambda x: len(token_lists[x]))\n",
    "df[\"unique_words\"] = df[\"species\"].map(lambda x: len(set(token_lists[x])))\n",
    "df[\"unique_stems\"] = df[\"species\"].map(lambda x: len(set(stems_lists[x])))\n",
    "df[\"total_lemmas\"] = df[\"species\"].map(lambda x: len(lemma_lists[x]))\n",
    "df[\"unique_lemmas\"] = df[\"species\"].map(lambda x: len(set(lemma_lists[x])))\n",
    "df[\"unique_lemmas_to_species\"] = df[\"species\"].map(lambda x: len(lemma_sets_unique_to_species[x]))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>num_words</th>\n",
       "      <th>num_sents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zma</td>\n",
       "      <td>159</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zma</td>\n",
       "      <td>54</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species  num_words  num_sents\n",
       "0      zma        159         20\n",
       "1      zma         54          6\n",
       "2      zma          3          1\n",
       "3      zma          3          1\n",
       "4      zma          3          1\n",
       "5      zma          3          1\n",
       "6      zma          3          1\n",
       "7      zma          3          1\n",
       "8      zma          3          1\n",
       "9      zma          3          1\n",
       "10     zma          3          1\n",
       "11     zma          3          1\n",
       "12     zma          3          1\n",
       "13     zma          3          1\n",
       "14     zma          3          1\n",
       "15     zma          3          1\n",
       "16     zma          3          1\n",
       "17     zma          3          1\n",
       "18     zma          3          1\n",
       "19     zma          3          1"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text_distributions = pd.DataFrame(dists)\n",
    "text_distributions.to_csv(os.path.join(OUTPUT_DIR, \"word_sent_distributions.csv\"), index=False)\n",
    "text_distributions.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>6</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>2</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ath</th>\n",
       "      <td>24006</td>\n",
       "      <td>17618</td>\n",
       "      <td>29181</td>\n",
       "      <td>60240</td>\n",
       "      <td>21522</td>\n",
       "      <td>29568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gmx</th>\n",
       "      <td>14</td>\n",
       "      <td>86</td>\n",
       "      <td>15</td>\n",
       "      <td>49</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mtr</th>\n",
       "      <td>296</td>\n",
       "      <td>178</td>\n",
       "      <td>307</td>\n",
       "      <td>443</td>\n",
       "      <td>250</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>osa</th>\n",
       "      <td>457</td>\n",
       "      <td>319</td>\n",
       "      <td>558</td>\n",
       "      <td>742</td>\n",
       "      <td>325</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sly</th>\n",
       "      <td>244</td>\n",
       "      <td>136</td>\n",
       "      <td>260</td>\n",
       "      <td>324</td>\n",
       "      <td>224</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zma</th>\n",
       "      <td>5987</td>\n",
       "      <td>1777</td>\n",
       "      <td>5481</td>\n",
       "      <td>7158</td>\n",
       "      <td>7739</td>\n",
       "      <td>3942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         3      6      4      5      2      1\n",
       "ath  24006  17618  29181  60240  21522  29568\n",
       "gmx     14     86     15     49     28     15\n",
       "mtr    296    178    307    443    250    217\n",
       "osa    457    319    558    742    325    190\n",
       "sly    244    136    260    324    224    150\n",
       "zma   5987   1777   5481   7158   7739   3942"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "species_to_num_to_quantity = {}\n",
    "for species in data.get_species():\n",
    "    how_many_species = lambda token: sum([(token in stems_lists[s]) for s in data.get_species()])\n",
    "    this_vocab = [token for token in stems_lists[species]]\n",
    "    distribution = [how_many_species(token) for token in this_vocab]\n",
    "    species_to_num_to_quantity[species] = dict(Counter(distribution))\n",
    "table = pd.DataFrame(species_to_num_to_quantity).transpose()\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>3</th>\n",
       "      <th>6</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>2</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>24006</td>\n",
       "      <td>17618</td>\n",
       "      <td>29181</td>\n",
       "      <td>60240</td>\n",
       "      <td>21522</td>\n",
       "      <td>29568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>14</td>\n",
       "      <td>86</td>\n",
       "      <td>15</td>\n",
       "      <td>49</td>\n",
       "      <td>28</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>296</td>\n",
       "      <td>178</td>\n",
       "      <td>307</td>\n",
       "      <td>443</td>\n",
       "      <td>250</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>457</td>\n",
       "      <td>319</td>\n",
       "      <td>558</td>\n",
       "      <td>742</td>\n",
       "      <td>325</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>244</td>\n",
       "      <td>136</td>\n",
       "      <td>260</td>\n",
       "      <td>324</td>\n",
       "      <td>224</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>5987</td>\n",
       "      <td>1777</td>\n",
       "      <td>5481</td>\n",
       "      <td>7158</td>\n",
       "      <td>7739</td>\n",
       "      <td>3942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species      3      6      4      5      2      1\n",
       "0     ath  24006  17618  29181  60240  21522  29568\n",
       "1     gmx     14     86     15     49     28     15\n",
       "2     mtr    296    178    307    443    250    217\n",
       "3     osa    457    319    558    742    325    190\n",
       "4     sly    244    136    260    324    224    150\n",
       "5     zma   5987   1777   5481   7158   7739   3942"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.DataFrame(species_to_num_to_quantity).transpose().reset_index()\n",
    "table.rename({\"index\":\"species\"}, axis=\"columns\", inplace=True)\n",
    "table.to_csv(os.path.join(OUTPUT_DIR, \"words_shared_by_species.csv\"), index=False)\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>others</th>\n",
       "      <th>quantity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>osa</td>\n",
       "      <td>5</td>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gmx</td>\n",
       "      <td>5</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>zma</td>\n",
       "      <td>5</td>\n",
       "      <td>1777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sly</td>\n",
       "      <td>5</td>\n",
       "      <td>136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>mtr</td>\n",
       "      <td>5</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ath</td>\n",
       "      <td>5</td>\n",
       "      <td>17618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>gmx</td>\n",
       "      <td>4</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ath</td>\n",
       "      <td>4</td>\n",
       "      <td>60240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>mtr</td>\n",
       "      <td>4</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>osa</td>\n",
       "      <td>4</td>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>sly</td>\n",
       "      <td>4</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>zma</td>\n",
       "      <td>4</td>\n",
       "      <td>7158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ath</td>\n",
       "      <td>3</td>\n",
       "      <td>29181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>gmx</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>mtr</td>\n",
       "      <td>3</td>\n",
       "      <td>307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>osa</td>\n",
       "      <td>3</td>\n",
       "      <td>558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>sly</td>\n",
       "      <td>3</td>\n",
       "      <td>260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>zma</td>\n",
       "      <td>3</td>\n",
       "      <td>5481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>2</td>\n",
       "      <td>24006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>2</td>\n",
       "      <td>5987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>2</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>2</td>\n",
       "      <td>457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>2</td>\n",
       "      <td>296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>zma</td>\n",
       "      <td>1</td>\n",
       "      <td>7739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>sly</td>\n",
       "      <td>1</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>osa</td>\n",
       "      <td>1</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>mtr</td>\n",
       "      <td>1</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>gmx</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ath</td>\n",
       "      <td>1</td>\n",
       "      <td>21522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ath</td>\n",
       "      <td>0</td>\n",
       "      <td>29568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>gmx</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>mtr</td>\n",
       "      <td>0</td>\n",
       "      <td>217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>osa</td>\n",
       "      <td>0</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>sly</td>\n",
       "      <td>0</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>zma</td>\n",
       "      <td>0</td>\n",
       "      <td>3942</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   species others  quantity\n",
       "9      osa      5       319\n",
       "7      gmx      5        86\n",
       "11     zma      5      1777\n",
       "10     sly      5       136\n",
       "8      mtr      5       178\n",
       "6      ath      5     17618\n",
       "19     gmx      4        49\n",
       "18     ath      4     60240\n",
       "20     mtr      4       443\n",
       "21     osa      4       742\n",
       "22     sly      4       324\n",
       "23     zma      4      7158\n",
       "12     ath      3     29181\n",
       "13     gmx      3        15\n",
       "14     mtr      3       307\n",
       "15     osa      3       558\n",
       "16     sly      3       260\n",
       "17     zma      3      5481\n",
       "0      ath      2     24006\n",
       "1      gmx      2        14\n",
       "5      zma      2      5987\n",
       "4      sly      2       244\n",
       "3      osa      2       457\n",
       "2      mtr      2       296\n",
       "29     zma      1      7739\n",
       "28     sly      1       224\n",
       "27     osa      1       325\n",
       "26     mtr      1       250\n",
       "25     gmx      1        28\n",
       "24     ath      1     21522\n",
       "30     ath      0     29568\n",
       "31     gmx      0        15\n",
       "32     mtr      0       217\n",
       "33     osa      0       190\n",
       "34     sly      0       150\n",
       "35     zma      0      3942"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniqueness_df = table.melt(id_vars=[\"species\"], var_name=\"others\", value_name=\"quantity\")\n",
    "uniqueness_df[\"others\"] = uniqueness_df[\"others\"]-1\n",
    "uniqueness_df.sort_values(by=\"others\", inplace=True, ascending=False)\n",
    "uniqueness_df.to_csv(os.path.join(OUTPUT_DIR, \"words_shared_by_species_melted.csv\"), index=False)\n",
    "uniqueness_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the ontology term annotations for each species?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6274</td>\n",
       "      <td>3818</td>\n",
       "      <td>30121</td>\n",
       "      <td>261441</td>\n",
       "      <td>7331</td>\n",
       "      <td>5305</td>\n",
       "      <td>261441</td>\n",
       "      <td>6792</td>\n",
       "      <td>5084</td>\n",
       "      <td>5973</td>\n",
       "      <td>4395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>7512</td>\n",
       "      <td>47139</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>47139</td>\n",
       "      <td>1722</td>\n",
       "      <td>498</td>\n",
       "      <td>170</td>\n",
       "      <td>761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>123</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7907</td>\n",
       "      <td>4841</td>\n",
       "      <td>38795</td>\n",
       "      <td>316616</td>\n",
       "      <td>8277</td>\n",
       "      <td>5984</td>\n",
       "      <td>316616</td>\n",
       "      <td>7663</td>\n",
       "      <td>5913</td>\n",
       "      <td>6149</td>\n",
       "      <td>5186</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po\n",
       "0     ath                     6274                 3818        30121       261441          7331          5305        261441           6792                      5084  5973  4395\n",
       "5     zma                     1405                  810         7512        47139          1846          1317         47139           1722                       498   170   761\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        97     0     0\n",
       "4     sly                       69                   69          359         1678           577           438          1678            552                        99     5     4\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       123     1     0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12     0    26\n",
       "6   total                     7907                 4841        38795       316616          8277          5984        316616           7663                      5913  6149  5186"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many of the genes in this dataset for each species are mapped to at least one term from a given ontology?\n",
    "num_mapped_go = {}\n",
    "num_mapped_po = {}\n",
    "for species in data.get_species():\n",
    "    d = data.to_pandas()\n",
    "    subset = d[d[\"species\"]==species]    \n",
    "    num_mapped_po[species] = len([t for t in subset[\"annotations\"].values if \"PO\" in t])\n",
    "    num_mapped_go[species] = len([t for t in subset[\"annotations\"].values if \"GO\" in t])\n",
    "num_mapped_go[\"total\"] = sum(list(num_mapped_go.values()))   \n",
    "num_mapped_po[\"total\"] = sum(list(num_mapped_po.values()))\n",
    "df[\"go\"] = df[\"species\"].map(lambda x: num_mapped_go[x])\n",
    "df[\"po\"] = df[\"species\"].map(lambda x: num_mapped_po[x])\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the biologically relevant groups like biochemical pathways and phenotypes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6274</td>\n",
       "      <td>3818</td>\n",
       "      <td>30121</td>\n",
       "      <td>261441</td>\n",
       "      <td>7331</td>\n",
       "      <td>5305</td>\n",
       "      <td>261441</td>\n",
       "      <td>6792</td>\n",
       "      <td>5084</td>\n",
       "      <td>5973</td>\n",
       "      <td>4395</td>\n",
       "      <td>1447</td>\n",
       "      <td>843</td>\n",
       "      <td>2356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>7512</td>\n",
       "      <td>47139</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>47139</td>\n",
       "      <td>1722</td>\n",
       "      <td>498</td>\n",
       "      <td>170</td>\n",
       "      <td>761</td>\n",
       "      <td>164</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>123</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7907</td>\n",
       "      <td>4841</td>\n",
       "      <td>38795</td>\n",
       "      <td>316616</td>\n",
       "      <td>8277</td>\n",
       "      <td>5984</td>\n",
       "      <td>316616</td>\n",
       "      <td>7663</td>\n",
       "      <td>5913</td>\n",
       "      <td>6149</td>\n",
       "      <td>5186</td>\n",
       "      <td>1631</td>\n",
       "      <td>983</td>\n",
       "      <td>2356</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke\n",
       "0     ath                     6274                 3818        30121       261441          7331          5305        261441           6792                      5084  5973  4395  1447       843          2356\n",
       "5     zma                     1405                  810         7512        47139          1846          1317         47139           1722                       498   170   761   164       133             0\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        97     0     0     0         3             0\n",
       "4     sly                       69                   69          359         1678           577           438          1678            552                        99     5     4    19         2             0\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       123     1     0     0         2             0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12     0    26     1         0             0\n",
       "6   total                     7907                 4841        38795       316616          8277          5984        316616           7663                      5913  6149  5186  1631       983          2356"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the groupings that we're interested in mapping to? Uses the paths defined at the top of the notebook.\n",
    "groupings_dict = {\n",
    "    \"kegg\":(kegg_pathways_path, kegg_pathways_names_path),\n",
    "    \"plantcyc\":(plantcyc_pathways_path, plantcyc_pathways_names_path),\n",
    "    \"lloyd_meinke\":(lloyd_meinke_subsets_path, lloyd_meinke_subsets_names_path)\n",
    "}\n",
    "\n",
    "\n",
    "for name,(filename,mapfile) in groupings_dict.items():\n",
    "    groups = Groupings(filename, {row.group_id:row.group_name for row in pd.read_csv(mapfile).itertuples()})\n",
    "    id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "    group_mapped_ids = [k for (k,v) in id_to_group_ids.items() if len(v)>0]\n",
    "    species_dict = data.get_species_dictionary()\n",
    "    num_mapped = {}\n",
    "    for species in data.get_species():\n",
    "        num_mapped[species] = len([x for x in group_mapped_ids if species_dict[x]==species])\n",
    "    num_mapped[\"total\"] = sum(list(num_mapped.values()))    \n",
    "    df[name] = df[\"species\"].map(lambda x: num_mapped[x])  \n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What about the other biologically relevant information like orthologous genes and protein interactions?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "      <th>panther</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6274</td>\n",
       "      <td>3818</td>\n",
       "      <td>30121</td>\n",
       "      <td>261441</td>\n",
       "      <td>7331</td>\n",
       "      <td>5305</td>\n",
       "      <td>261441</td>\n",
       "      <td>6792</td>\n",
       "      <td>5084</td>\n",
       "      <td>5973</td>\n",
       "      <td>4395</td>\n",
       "      <td>1447</td>\n",
       "      <td>843</td>\n",
       "      <td>2356</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>7512</td>\n",
       "      <td>47139</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>47139</td>\n",
       "      <td>1722</td>\n",
       "      <td>498</td>\n",
       "      <td>170</td>\n",
       "      <td>761</td>\n",
       "      <td>164</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>123</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7907</td>\n",
       "      <td>4841</td>\n",
       "      <td>38795</td>\n",
       "      <td>316616</td>\n",
       "      <td>8277</td>\n",
       "      <td>5984</td>\n",
       "      <td>316616</td>\n",
       "      <td>7663</td>\n",
       "      <td>5913</td>\n",
       "      <td>6149</td>\n",
       "      <td>5186</td>\n",
       "      <td>1631</td>\n",
       "      <td>983</td>\n",
       "      <td>2356</td>\n",
       "      <td>916</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke  panther\n",
       "0     ath                     6274                 3818        30121       261441          7331          5305        261441           6792                      5084  5973  4395  1447       843          2356      377\n",
       "5     zma                     1405                  810         7512        47139          1846          1317         47139           1722                       498   170   761   164       133             0      443\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        97     0     0     0         3             0       86\n",
       "4     sly                       69                   69          359         1678           577           438          1678            552                        99     5     4    19         2             0       10\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       123     1     0     0         2             0        0\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12     0    26     1         0             0        0\n",
       "6   total                     7907                 4841        38795       316616          8277          5984        316616           7663                      5913  6149  5186  1631       983          2356      916"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PantherDB for plant orthologs.\n",
    "ortholog_edgelist = AnyInteractions(data.get_name_to_id_dictionary(), ortholog_file_path)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in ortholog_edgelist.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"panther\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>unique_gene_identifiers</th>\n",
       "      <th>unique_descriptions</th>\n",
       "      <th>total_sents</th>\n",
       "      <th>total_words</th>\n",
       "      <th>unique_words</th>\n",
       "      <th>unique_stems</th>\n",
       "      <th>total_lemmas</th>\n",
       "      <th>unique_lemmas</th>\n",
       "      <th>unique_lemmas_to_species</th>\n",
       "      <th>go</th>\n",
       "      <th>po</th>\n",
       "      <th>kegg</th>\n",
       "      <th>plantcyc</th>\n",
       "      <th>lloyd_meinke</th>\n",
       "      <th>panther</th>\n",
       "      <th>stringdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ath</td>\n",
       "      <td>6274</td>\n",
       "      <td>3818</td>\n",
       "      <td>30121</td>\n",
       "      <td>261441</td>\n",
       "      <td>7331</td>\n",
       "      <td>5305</td>\n",
       "      <td>261441</td>\n",
       "      <td>6792</td>\n",
       "      <td>5084</td>\n",
       "      <td>5973</td>\n",
       "      <td>4395</td>\n",
       "      <td>1447</td>\n",
       "      <td>843</td>\n",
       "      <td>2356</td>\n",
       "      <td>377</td>\n",
       "      <td>4317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>zma</td>\n",
       "      <td>1405</td>\n",
       "      <td>810</td>\n",
       "      <td>7512</td>\n",
       "      <td>47139</td>\n",
       "      <td>1846</td>\n",
       "      <td>1317</td>\n",
       "      <td>47139</td>\n",
       "      <td>1722</td>\n",
       "      <td>498</td>\n",
       "      <td>170</td>\n",
       "      <td>761</td>\n",
       "      <td>164</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "      <td>443</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>osa</td>\n",
       "      <td>92</td>\n",
       "      <td>85</td>\n",
       "      <td>478</td>\n",
       "      <td>3689</td>\n",
       "      <td>826</td>\n",
       "      <td>586</td>\n",
       "      <td>3689</td>\n",
       "      <td>760</td>\n",
       "      <td>97</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>86</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sly</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>359</td>\n",
       "      <td>1678</td>\n",
       "      <td>577</td>\n",
       "      <td>438</td>\n",
       "      <td>1678</td>\n",
       "      <td>552</td>\n",
       "      <td>99</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mtr</td>\n",
       "      <td>37</td>\n",
       "      <td>36</td>\n",
       "      <td>263</td>\n",
       "      <td>2447</td>\n",
       "      <td>718</td>\n",
       "      <td>516</td>\n",
       "      <td>2447</td>\n",
       "      <td>671</td>\n",
       "      <td>123</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gmx</td>\n",
       "      <td>30</td>\n",
       "      <td>23</td>\n",
       "      <td>62</td>\n",
       "      <td>222</td>\n",
       "      <td>81</td>\n",
       "      <td>68</td>\n",
       "      <td>222</td>\n",
       "      <td>78</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>total</td>\n",
       "      <td>7907</td>\n",
       "      <td>4841</td>\n",
       "      <td>38795</td>\n",
       "      <td>316616</td>\n",
       "      <td>8277</td>\n",
       "      <td>5984</td>\n",
       "      <td>316616</td>\n",
       "      <td>7663</td>\n",
       "      <td>5913</td>\n",
       "      <td>6149</td>\n",
       "      <td>5186</td>\n",
       "      <td>1631</td>\n",
       "      <td>983</td>\n",
       "      <td>2356</td>\n",
       "      <td>916</td>\n",
       "      <td>4620</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  species  unique_gene_identifiers  unique_descriptions  total_sents  total_words  unique_words  unique_stems  total_lemmas  unique_lemmas  unique_lemmas_to_species    go    po  kegg  plantcyc  lloyd_meinke  panther  stringdb\n",
       "0     ath                     6274                 3818        30121       261441          7331          5305        261441           6792                      5084  5973  4395  1447       843          2356      377      4317\n",
       "5     zma                     1405                  810         7512        47139          1846          1317         47139           1722                       498   170   761   164       133             0      443       229\n",
       "3     osa                       92                   85          478         3689           826           586          3689            760                        97     0     0     0         3             0       86        45\n",
       "4     sly                       69                   69          359         1678           577           438          1678            552                        99     5     4    19         2             0       10        11\n",
       "2     mtr                       37                   36          263         2447           718           516          2447            671                       123     1     0     0         2             0        0        13\n",
       "1     gmx                       30                   23           62          222            81            68           222             78                        12     0    26     1         0             0        0         5\n",
       "6   total                     7907                 4841        38795       316616          8277          5984        316616           7663                      5913  6149  5186  1631       983          2356      916      4620"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# STRING DB for protein-protein interactions.\n",
    "naming_file = \"../../plant-data/databases/string/all_organisms.name_2_string.tsv\"\n",
    "interaction_files = [\n",
    "    \"../../plant-data/databases/string/3702.protein.links.detailed.v11.0.txt\", # Arabidopsis\n",
    "    \"../../plant-data/databases/string/4577.protein.links.detailed.v11.0.txt\", # Maize\n",
    "    \"../../plant-data/databases/string/4530.protein.links.detailed.v11.0.txt\", # Tomato \n",
    "    \"../../plant-data/databases/string/4081.protein.links.detailed.v11.0.txt\", # Medicago\n",
    "    \"../../plant-data/databases/string/3880.protein.links.detailed.v11.0.txt\", # Rice \n",
    "    \"../../plant-data/databases/string/3847.protein.links.detailed.v11.0.txt\", # Soybean\n",
    "    \"../../plant-data/databases/string/9606.protein.links.detailed.v11.0.txt\", # Human\n",
    "]\n",
    "genes = data.get_gene_dictionary()\n",
    "string_data = ProteinInteractions(genes, naming_file, *interaction_files)\n",
    "species_dict = data.get_species_dictionary()\n",
    "num_mapped = {}\n",
    "for species in data.get_species():\n",
    "    num_mapped[species] = len([x for x in string_data.ids if species_dict[x]==species])\n",
    "num_mapped[\"total\"] = sum(list(num_mapped.values()))\n",
    "df[\"stringdb\"] = df[\"species\"].map(lambda x: num_mapped[x])    \n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write that dataframe with all the information about dataset to a file.\n",
    "df.to_csv(os.path.join(OUTPUT_DIR,\"full_dataset_composition.csv\"),index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### How do the vocabularies used for different species compare?\n",
    "One of the things we are interested in is discovering or recovering phenotype similarity between different species in order to identify phenologs (phenotypes between species that share some underlying genetic cause). For this reason, we are interested in how the vocabularies used to describe phenotypes between different species vary, because this will impact how feasible it is to use a dataset like this to identify phenologs. Because the Arabidopsis and maize datasets are the largest in this case, we will compare the vocabularies used in describing the phenotypes associated with the genes from these species in this dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mutation</td>\n",
       "      <td>NN</td>\n",
       "      <td>210</td>\n",
       "      <td>0.080324</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002121</td>\n",
       "      <td>0.078203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ibr</td>\n",
       "      <td>NN</td>\n",
       "      <td>11</td>\n",
       "      <td>0.004207</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>represses</td>\n",
       "      <td>NN</td>\n",
       "      <td>10</td>\n",
       "      <td>0.003825</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>proteasome</td>\n",
       "      <td>NN</td>\n",
       "      <td>19</td>\n",
       "      <td>0.007267</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mildly</td>\n",
       "      <td>RB</td>\n",
       "      <td>14</td>\n",
       "      <td>0.005355</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>speciation</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>subpopulation</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nutritional</td>\n",
       "      <td>JJ</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pc</td>\n",
       "      <td>NN</td>\n",
       "      <td>7</td>\n",
       "      <td>0.002677</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>mildy</td>\n",
       "      <td>NN</td>\n",
       "      <td>4</td>\n",
       "      <td>0.001530</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001530</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "0       mutation             NN       210  0.080324         1  0.002121  0.078203\n",
       "1            ibr             NN        11  0.004207         0  0.000000  0.004207\n",
       "2      represses             NN        10  0.003825         0  0.000000  0.003825\n",
       "3     proteasome             NN        19  0.007267         0  0.000000  0.007267\n",
       "4         mildly             RB        14  0.005355         0  0.000000  0.005355\n",
       "5     speciation             NN         1  0.000382         0  0.000000  0.000382\n",
       "6  subpopulation             NN         1  0.000382         0  0.000000  0.000382\n",
       "7    nutritional             JJ         1  0.000382         0  0.000000  0.000382\n",
       "8             pc             NN         7  0.002677         0  0.000000  0.002677\n",
       "9          mildy             NN         4  0.001530         0  0.000000  0.001530"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Using lemmas as the vocabulary components.\n",
    "vocabs = {s:set(lemma_list) for s,lemma_list in lemma_lists.items()}\n",
    "fdist_zma = FreqDist(lemma_lists[\"zma\"])\n",
    "fdist_ath = FreqDist(lemma_lists[\"ath\"])\n",
    "\n",
    "# Using word stems as the vocabulary components.\n",
    "#vocabs = {s:set(stems_list) for s,stems_list in stems_lists.items()}\n",
    "#fdist_zma = FreqDist(stems_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(stems_lists[\"ath\"])\n",
    "\n",
    "# Using tokens (full words) as the vocabulary components.\n",
    "#vocabs = {s:set(token_list) for s,token_list in token_lists.items()}\n",
    "#fdist_zma = FreqDist(token_lists[\"zma\"])\n",
    "#fdist_ath = FreqDist(token_lists[\"ath\"])\n",
    "\n",
    "union_vocab = vocabs[\"zma\"].union(vocabs[\"ath\"])\n",
    "table = pd.DataFrame({\"token\":list(union_vocab)})\n",
    "stops = set(stopwords.words('english'))\n",
    "table = table[~table.token.isin(stops)]\n",
    "table[\"part_of_speech\"] = table[\"token\"].map(lambda x: nltk.pos_tag([x])[0][1][:2])\n",
    "table[\"ath_freq\"] = table[\"token\"].map(lambda x: fdist_ath[x])\n",
    "table[\"ath_rate\"] = table[\"ath_freq\"]*100/len(token_lists[\"ath\"])\n",
    "table[\"zma_freq\"] = table[\"token\"].map(lambda x: fdist_zma[x])\n",
    "table[\"zma_rate\"] = table[\"zma_freq\"]*100/len(token_lists[\"zma\"])\n",
    "table[\"diff\"] = table[\"ath_rate\"]-table[\"zma_rate\"]\n",
    "table.to_csv(os.path.join(OUTPUT_DIR,\"token_frequencies.csv\"), index=False)\n",
    "table.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>embryo</td>\n",
       "      <td>NN</td>\n",
       "      <td>4739</td>\n",
       "      <td>1.812646</td>\n",
       "      <td>149</td>\n",
       "      <td>0.316086</td>\n",
       "      <td>1.496560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5784</th>\n",
       "      <td>mutant</td>\n",
       "      <td>NN</td>\n",
       "      <td>4870</td>\n",
       "      <td>1.862753</td>\n",
       "      <td>254</td>\n",
       "      <td>0.538832</td>\n",
       "      <td>1.323921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1134</th>\n",
       "      <td>phenotype</td>\n",
       "      <td>NN</td>\n",
       "      <td>3647</td>\n",
       "      <td>1.394961</td>\n",
       "      <td>76</td>\n",
       "      <td>0.161225</td>\n",
       "      <td>1.233736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3317</th>\n",
       "      <td>type</td>\n",
       "      <td>NN</td>\n",
       "      <td>2562</td>\n",
       "      <td>0.979953</td>\n",
       "      <td>15</td>\n",
       "      <td>0.031821</td>\n",
       "      <td>0.948133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4849</th>\n",
       "      <td>root</td>\n",
       "      <td>NN</td>\n",
       "      <td>2785</td>\n",
       "      <td>1.065250</td>\n",
       "      <td>64</td>\n",
       "      <td>0.135769</td>\n",
       "      <td>0.929481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5433</th>\n",
       "      <td>wild</td>\n",
       "      <td>NN</td>\n",
       "      <td>2465</td>\n",
       "      <td>0.942851</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014850</td>\n",
       "      <td>0.928002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6808</th>\n",
       "      <td>reduced</td>\n",
       "      <td>VB</td>\n",
       "      <td>2856</td>\n",
       "      <td>1.092407</td>\n",
       "      <td>198</td>\n",
       "      <td>0.420034</td>\n",
       "      <td>0.672373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4075</th>\n",
       "      <td>stage</td>\n",
       "      <td>NN</td>\n",
       "      <td>1812</td>\n",
       "      <td>0.693082</td>\n",
       "      <td>54</td>\n",
       "      <td>0.114555</td>\n",
       "      <td>0.578527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3495</th>\n",
       "      <td>growth</td>\n",
       "      <td>NN</td>\n",
       "      <td>1827</td>\n",
       "      <td>0.698819</td>\n",
       "      <td>67</td>\n",
       "      <td>0.142133</td>\n",
       "      <td>0.556686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3717</th>\n",
       "      <td>cotyledon</td>\n",
       "      <td>NN</td>\n",
       "      <td>1383</td>\n",
       "      <td>0.528991</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.528991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5885</th>\n",
       "      <td>defective</td>\n",
       "      <td>JJ</td>\n",
       "      <td>2937</td>\n",
       "      <td>1.123389</td>\n",
       "      <td>281</td>\n",
       "      <td>0.596109</td>\n",
       "      <td>0.527280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6154</th>\n",
       "      <td>seed</td>\n",
       "      <td>NN</td>\n",
       "      <td>1874</td>\n",
       "      <td>0.716797</td>\n",
       "      <td>103</td>\n",
       "      <td>0.218503</td>\n",
       "      <td>0.498294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4417</th>\n",
       "      <td>terminal</td>\n",
       "      <td>NN</td>\n",
       "      <td>1265</td>\n",
       "      <td>0.483857</td>\n",
       "      <td>3</td>\n",
       "      <td>0.006364</td>\n",
       "      <td>0.477493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1111</th>\n",
       "      <td>increased</td>\n",
       "      <td>VB</td>\n",
       "      <td>1299</td>\n",
       "      <td>0.496862</td>\n",
       "      <td>11</td>\n",
       "      <td>0.023335</td>\n",
       "      <td>0.473526</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5486</th>\n",
       "      <td>level</td>\n",
       "      <td>NN</td>\n",
       "      <td>1327</td>\n",
       "      <td>0.507571</td>\n",
       "      <td>19</td>\n",
       "      <td>0.040306</td>\n",
       "      <td>0.467265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6624</th>\n",
       "      <td>cell</td>\n",
       "      <td>NN</td>\n",
       "      <td>1711</td>\n",
       "      <td>0.654450</td>\n",
       "      <td>97</td>\n",
       "      <td>0.205774</td>\n",
       "      <td>0.448675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6270</th>\n",
       "      <td>double</td>\n",
       "      <td>RB</td>\n",
       "      <td>981</td>\n",
       "      <td>0.375228</td>\n",
       "      <td>3</td>\n",
       "      <td>0.006364</td>\n",
       "      <td>0.368864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6348</th>\n",
       "      <td>globular</td>\n",
       "      <td>NN</td>\n",
       "      <td>896</td>\n",
       "      <td>0.342716</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.342716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2702</th>\n",
       "      <td>wa</td>\n",
       "      <td>NN</td>\n",
       "      <td>897</td>\n",
       "      <td>0.343098</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002121</td>\n",
       "      <td>0.340977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>rosette</td>\n",
       "      <td>NN</td>\n",
       "      <td>892</td>\n",
       "      <td>0.341186</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002121</td>\n",
       "      <td>0.339065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2317</th>\n",
       "      <td>compared</td>\n",
       "      <td>VB</td>\n",
       "      <td>933</td>\n",
       "      <td>0.356868</td>\n",
       "      <td>9</td>\n",
       "      <td>0.019092</td>\n",
       "      <td>0.337776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2462</th>\n",
       "      <td>flower</td>\n",
       "      <td>NN</td>\n",
       "      <td>843</td>\n",
       "      <td>0.322444</td>\n",
       "      <td>7</td>\n",
       "      <td>0.014850</td>\n",
       "      <td>0.307594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3468</th>\n",
       "      <td>visible</td>\n",
       "      <td>JJ</td>\n",
       "      <td>848</td>\n",
       "      <td>0.324356</td>\n",
       "      <td>11</td>\n",
       "      <td>0.023335</td>\n",
       "      <td>0.301021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5447</th>\n",
       "      <td>silique</td>\n",
       "      <td>NN</td>\n",
       "      <td>777</td>\n",
       "      <td>0.297199</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.297199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3498</th>\n",
       "      <td>development</td>\n",
       "      <td>NN</td>\n",
       "      <td>859</td>\n",
       "      <td>0.328564</td>\n",
       "      <td>19</td>\n",
       "      <td>0.040306</td>\n",
       "      <td>0.288257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5903</th>\n",
       "      <td>abnormal</td>\n",
       "      <td>JJ</td>\n",
       "      <td>894</td>\n",
       "      <td>0.341951</td>\n",
       "      <td>26</td>\n",
       "      <td>0.055156</td>\n",
       "      <td>0.286795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3991</th>\n",
       "      <td>hypocotyl</td>\n",
       "      <td>NN</td>\n",
       "      <td>743</td>\n",
       "      <td>0.284194</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.284194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4421</th>\n",
       "      <td>defect</td>\n",
       "      <td>NN</td>\n",
       "      <td>843</td>\n",
       "      <td>0.322444</td>\n",
       "      <td>19</td>\n",
       "      <td>0.040306</td>\n",
       "      <td>0.282137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3766</th>\n",
       "      <td>sensitive</td>\n",
       "      <td>JJ</td>\n",
       "      <td>761</td>\n",
       "      <td>0.291079</td>\n",
       "      <td>6</td>\n",
       "      <td>0.012728</td>\n",
       "      <td>0.278351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4027</th>\n",
       "      <td>flowering</td>\n",
       "      <td>VB</td>\n",
       "      <td>899</td>\n",
       "      <td>0.343863</td>\n",
       "      <td>31</td>\n",
       "      <td>0.065763</td>\n",
       "      <td>0.278100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "41         embryo             NN      4739  1.812646       149  0.316086  1.496560\n",
       "5784       mutant             NN      4870  1.862753       254  0.538832  1.323921\n",
       "1134    phenotype             NN      3647  1.394961        76  0.161225  1.233736\n",
       "3317         type             NN      2562  0.979953        15  0.031821  0.948133\n",
       "4849         root             NN      2785  1.065250        64  0.135769  0.929481\n",
       "5433         wild             NN      2465  0.942851         7  0.014850  0.928002\n",
       "6808      reduced             VB      2856  1.092407       198  0.420034  0.672373\n",
       "4075        stage             NN      1812  0.693082        54  0.114555  0.578527\n",
       "3495       growth             NN      1827  0.698819        67  0.142133  0.556686\n",
       "3717    cotyledon             NN      1383  0.528991         0  0.000000  0.528991\n",
       "5885    defective             JJ      2937  1.123389       281  0.596109  0.527280\n",
       "6154         seed             NN      1874  0.716797       103  0.218503  0.498294\n",
       "4417     terminal             NN      1265  0.483857         3  0.006364  0.477493\n",
       "1111    increased             VB      1299  0.496862        11  0.023335  0.473526\n",
       "5486        level             NN      1327  0.507571        19  0.040306  0.467265\n",
       "6624         cell             NN      1711  0.654450        97  0.205774  0.448675\n",
       "6270       double             RB       981  0.375228         3  0.006364  0.368864\n",
       "6348     globular             NN       896  0.342716         0  0.000000  0.342716\n",
       "2702           wa             NN       897  0.343098         1  0.002121  0.340977\n",
       "596       rosette             NN       892  0.341186         1  0.002121  0.339065\n",
       "2317     compared             VB       933  0.356868         9  0.019092  0.337776\n",
       "2462       flower             NN       843  0.322444         7  0.014850  0.307594\n",
       "3468      visible             JJ       848  0.324356        11  0.023335  0.301021\n",
       "5447      silique             NN       777  0.297199         0  0.000000  0.297199\n",
       "3498  development             NN       859  0.328564        19  0.040306  0.288257\n",
       "5903     abnormal             JJ       894  0.341951        26  0.055156  0.286795\n",
       "3991    hypocotyl             NN       743  0.284194         0  0.000000  0.284194\n",
       "4421       defect             NN       843  0.322444        19  0.040306  0.282137\n",
       "3766    sensitive             JJ       761  0.291079         6  0.012728  0.278351\n",
       "4027    flowering             VB       899  0.343863        31  0.065763  0.278100"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for Arabidopsis than maize descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=False, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>ath_freq</th>\n",
       "      <th>ath_rate</th>\n",
       "      <th>zma_freq</th>\n",
       "      <th>zma_rate</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1043</th>\n",
       "      <td>endosperm</td>\n",
       "      <td>NN</td>\n",
       "      <td>120</td>\n",
       "      <td>0.045899</td>\n",
       "      <td>1033</td>\n",
       "      <td>2.191391</td>\n",
       "      <td>-2.145492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2415</th>\n",
       "      <td>seedling</td>\n",
       "      <td>VB</td>\n",
       "      <td>1457</td>\n",
       "      <td>0.557296</td>\n",
       "      <td>1221</td>\n",
       "      <td>2.590212</td>\n",
       "      <td>-2.032916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2178</th>\n",
       "      <td>yellow</td>\n",
       "      <td>NN</td>\n",
       "      <td>274</td>\n",
       "      <td>0.104804</td>\n",
       "      <td>761</td>\n",
       "      <td>1.614375</td>\n",
       "      <td>-1.509571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5556</th>\n",
       "      <td>kernel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>709</td>\n",
       "      <td>1.504062</td>\n",
       "      <td>-1.504062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3469</th>\n",
       "      <td>leaf</td>\n",
       "      <td>NN</td>\n",
       "      <td>3625</td>\n",
       "      <td>1.386546</td>\n",
       "      <td>1356</td>\n",
       "      <td>2.876599</td>\n",
       "      <td>-1.490053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2695</th>\n",
       "      <td>green</td>\n",
       "      <td>JJ</td>\n",
       "      <td>808</td>\n",
       "      <td>0.309056</td>\n",
       "      <td>773</td>\n",
       "      <td>1.639831</td>\n",
       "      <td>-1.330775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3032</th>\n",
       "      <td>white</td>\n",
       "      <td>JJ</td>\n",
       "      <td>336</td>\n",
       "      <td>0.128518</td>\n",
       "      <td>621</td>\n",
       "      <td>1.317381</td>\n",
       "      <td>-1.188862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4066</th>\n",
       "      <td>usually</td>\n",
       "      <td>RB</td>\n",
       "      <td>42</td>\n",
       "      <td>0.016065</td>\n",
       "      <td>336</td>\n",
       "      <td>0.712786</td>\n",
       "      <td>-0.696721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6127</th>\n",
       "      <td>color</td>\n",
       "      <td>NN</td>\n",
       "      <td>48</td>\n",
       "      <td>0.018360</td>\n",
       "      <td>297</td>\n",
       "      <td>0.630052</td>\n",
       "      <td>-0.611692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>albino</td>\n",
       "      <td>NN</td>\n",
       "      <td>195</td>\n",
       "      <td>0.074587</td>\n",
       "      <td>317</td>\n",
       "      <td>0.672479</td>\n",
       "      <td>-0.597893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2564</th>\n",
       "      <td>tassel</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>260</td>\n",
       "      <td>0.551560</td>\n",
       "      <td>-0.551560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3232</th>\n",
       "      <td>ear</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>252</td>\n",
       "      <td>0.534589</td>\n",
       "      <td>-0.534589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3387</th>\n",
       "      <td>often</td>\n",
       "      <td>RB</td>\n",
       "      <td>140</td>\n",
       "      <td>0.053549</td>\n",
       "      <td>269</td>\n",
       "      <td>0.570653</td>\n",
       "      <td>-0.517103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2831</th>\n",
       "      <td>class</td>\n",
       "      <td>NN</td>\n",
       "      <td>8</td>\n",
       "      <td>0.003060</td>\n",
       "      <td>240</td>\n",
       "      <td>0.509133</td>\n",
       "      <td>-0.506073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5169</th>\n",
       "      <td>opaque</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>236</td>\n",
       "      <td>0.500647</td>\n",
       "      <td>-0.500265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2320</th>\n",
       "      <td>aleurone</td>\n",
       "      <td>NN</td>\n",
       "      <td>3</td>\n",
       "      <td>0.001147</td>\n",
       "      <td>213</td>\n",
       "      <td>0.451855</td>\n",
       "      <td>-0.450708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6898</th>\n",
       "      <td>tissue</td>\n",
       "      <td>NN</td>\n",
       "      <td>234</td>\n",
       "      <td>0.089504</td>\n",
       "      <td>247</td>\n",
       "      <td>0.523982</td>\n",
       "      <td>-0.434478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2465</th>\n",
       "      <td>small</td>\n",
       "      <td>JJ</td>\n",
       "      <td>649</td>\n",
       "      <td>0.248240</td>\n",
       "      <td>318</td>\n",
       "      <td>0.674601</td>\n",
       "      <td>-0.426361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6548</th>\n",
       "      <td>pale</td>\n",
       "      <td>NN</td>\n",
       "      <td>581</td>\n",
       "      <td>0.222230</td>\n",
       "      <td>305</td>\n",
       "      <td>0.647023</td>\n",
       "      <td>-0.424793</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6189</th>\n",
       "      <td>vp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>190</td>\n",
       "      <td>0.403063</td>\n",
       "      <td>-0.403063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>palegreen</td>\n",
       "      <td>NN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000382</td>\n",
       "      <td>171</td>\n",
       "      <td>0.362757</td>\n",
       "      <td>-0.362374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2301</th>\n",
       "      <td>texture</td>\n",
       "      <td>NN</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000765</td>\n",
       "      <td>156</td>\n",
       "      <td>0.330936</td>\n",
       "      <td>-0.330171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>surface</td>\n",
       "      <td>NN</td>\n",
       "      <td>143</td>\n",
       "      <td>0.054697</td>\n",
       "      <td>181</td>\n",
       "      <td>0.383971</td>\n",
       "      <td>-0.329274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3076</th>\n",
       "      <td>otherwise</td>\n",
       "      <td>RB</td>\n",
       "      <td>10</td>\n",
       "      <td>0.003825</td>\n",
       "      <td>157</td>\n",
       "      <td>0.333058</td>\n",
       "      <td>-0.329233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2373</th>\n",
       "      <td>general</td>\n",
       "      <td>JJ</td>\n",
       "      <td>14</td>\n",
       "      <td>0.005355</td>\n",
       "      <td>157</td>\n",
       "      <td>0.333058</td>\n",
       "      <td>-0.327703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3335</th>\n",
       "      <td>appearance</td>\n",
       "      <td>NN</td>\n",
       "      <td>133</td>\n",
       "      <td>0.050872</td>\n",
       "      <td>178</td>\n",
       "      <td>0.377607</td>\n",
       "      <td>-0.326735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1102</th>\n",
       "      <td>term</td>\n",
       "      <td>NN</td>\n",
       "      <td>28</td>\n",
       "      <td>0.010710</td>\n",
       "      <td>157</td>\n",
       "      <td>0.333058</td>\n",
       "      <td>-0.322348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>714</th>\n",
       "      <td>sometimes</td>\n",
       "      <td>RB</td>\n",
       "      <td>37</td>\n",
       "      <td>0.014152</td>\n",
       "      <td>142</td>\n",
       "      <td>0.301237</td>\n",
       "      <td>-0.287084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6858</th>\n",
       "      <td>pericarp</td>\n",
       "      <td>NN</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>132</td>\n",
       "      <td>0.280023</td>\n",
       "      <td>-0.280023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4554</th>\n",
       "      <td>necrotic</td>\n",
       "      <td>JJ</td>\n",
       "      <td>42</td>\n",
       "      <td>0.016065</td>\n",
       "      <td>138</td>\n",
       "      <td>0.292751</td>\n",
       "      <td>-0.276686</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           token part_of_speech  ath_freq  ath_rate  zma_freq  zma_rate      diff\n",
       "1043   endosperm             NN       120  0.045899      1033  2.191391 -2.145492\n",
       "2415    seedling             VB      1457  0.557296      1221  2.590212 -2.032916\n",
       "2178      yellow             NN       274  0.104804       761  1.614375 -1.509571\n",
       "5556      kernel             NN         0  0.000000       709  1.504062 -1.504062\n",
       "3469        leaf             NN      3625  1.386546      1356  2.876599 -1.490053\n",
       "2695       green             JJ       808  0.309056       773  1.639831 -1.330775\n",
       "3032       white             JJ       336  0.128518       621  1.317381 -1.188862\n",
       "4066     usually             RB        42  0.016065       336  0.712786 -0.696721\n",
       "6127       color             NN        48  0.018360       297  0.630052 -0.611692\n",
       "904       albino             NN       195  0.074587       317  0.672479 -0.597893\n",
       "2564      tassel             NN         0  0.000000       260  0.551560 -0.551560\n",
       "3232         ear             NN         0  0.000000       252  0.534589 -0.534589\n",
       "3387       often             RB       140  0.053549       269  0.570653 -0.517103\n",
       "2831       class             NN         8  0.003060       240  0.509133 -0.506073\n",
       "5169      opaque             NN         1  0.000382       236  0.500647 -0.500265\n",
       "2320    aleurone             NN         3  0.001147       213  0.451855 -0.450708\n",
       "6898      tissue             NN       234  0.089504       247  0.523982 -0.434478\n",
       "2465       small             JJ       649  0.248240       318  0.674601 -0.426361\n",
       "6548        pale             NN       581  0.222230       305  0.647023 -0.424793\n",
       "6189          vp             NN         0  0.000000       190  0.403063 -0.403063\n",
       "195    palegreen             NN         1  0.000382       171  0.362757 -0.362374\n",
       "2301     texture             NN         2  0.000765       156  0.330936 -0.330171\n",
       "160      surface             NN       143  0.054697       181  0.383971 -0.329274\n",
       "3076   otherwise             RB        10  0.003825       157  0.333058 -0.329233\n",
       "2373     general             JJ        14  0.005355       157  0.333058 -0.327703\n",
       "3335  appearance             NN       133  0.050872       178  0.377607 -0.326735\n",
       "1102        term             NN        28  0.010710       157  0.333058 -0.322348\n",
       "714    sometimes             RB        37  0.014152       142  0.301237 -0.287084\n",
       "6858    pericarp             NN         0  0.000000       132  0.280023 -0.280023\n",
       "4554    necrotic             JJ        42  0.016065       138  0.292751 -0.276686"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# What are the tokens more frequently used for maize than Arabidopsis descriptions in this dataset?\n",
    "table.sort_values(by=\"diff\", ascending=True, inplace=True)\n",
    "table.head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>part_of_speech</th>\n",
       "      <th>abs_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MD</td>\n",
       "      <td>0.061224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD</td>\n",
       "      <td>0.029343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IN</td>\n",
       "      <td>0.025702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JJ</td>\n",
       "      <td>0.022987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DT</td>\n",
       "      <td>0.020502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RB</td>\n",
       "      <td>0.016103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NN</td>\n",
       "      <td>0.013981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>VB</td>\n",
       "      <td>0.012056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CC</td>\n",
       "      <td>0.008103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WP</td>\n",
       "      <td>0.001739</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  part_of_speech  abs_diff\n",
       "0             MD  0.061224\n",
       "1             CD  0.029343\n",
       "2             IN  0.025702\n",
       "3             JJ  0.022987\n",
       "4             DT  0.020502\n",
       "5             RB  0.016103\n",
       "6             NN  0.013981\n",
       "7             VB  0.012056\n",
       "8             CC  0.008103\n",
       "9             WP  0.001739"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is the mean absolute value of the rate differences between the different parts of speech?\n",
    "table[\"abs_diff\"] = abs(table[\"diff\"])\n",
    "pos_table = table.groupby(\"part_of_speech\").mean()\n",
    "pos_table.sort_values(by=\"abs_diff\", inplace=True, ascending=False)\n",
    "pos_table = pos_table[[\"abs_diff\"]]\n",
    "pos_table.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working on the Venn Diagram for this part, unused currently.\n",
    "#print(table.shape)\n",
    "#zma_only = table[table[\"ath_rate\"]==0]\n",
    "#ath_only = table[table[\"zma_rate\"]==0]\n",
    "#print(zma_only.shape)\n",
    "#print(ath_only.shape)\n",
    "#print(ath_only.shape[0]+zma_only.shape[0])\n",
    "#ath_only.head(10)\n",
    "# We need to create a mapping between stems and the words that were present for them.\n",
    "# This is because what we want is the stems that are exclusive to a species.\n",
    "# but then the words that are actually there for those stems, so that we can count their parts of speech."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Looking at Term and Word Enrichment for Groups of Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5186"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Loading the dataset of phenotype descriptions and ontology annotations.\n",
    "plant_dataset = Dataset(plant_dataset_path)\n",
    "data = plant_dataset\n",
    "data.filter_has_description()\n",
    "#data.filter_has_annotation(\"GO\")\n",
    "data.filter_has_annotation(\"PO\")\n",
    "d = data.get_description_dictionary()\n",
    "texts = {i:\" \".join(simple_preprocess(t)) for i,t in d.items()}\n",
    "len(texts)                              "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create ontology objects for all the biological ontologies being used.\n",
    "#go_pickle_path = \"../ontologies/go.pickle\"                                                                \n",
    "#po_pickle_path = \"../ontologies/po.pickle\"                                                             \n",
    "#pato_pickle_path = \"../ontologies/pato.pickle\"\n",
    "\n",
    "#pato = load_from_pickle(pato_pickle_path)\n",
    "#po = load_from_pickle(po_pickle_path)\n",
    "#go = load_from_pickle(go_pickle_path)\n",
    "\n",
    "go_obo_path = \"../ontologies/go.obo\"                                                                \n",
    "po_obo_path = \"../ontologies/po.obo\"                                                             \n",
    "pato_obo_path = \"../ontologies/pato.obo\"\n",
    "\n",
    "po = Ontology(po_obo_path)\n",
    "go = Ontology(go_obo_path)\n",
    "pato = Ontology(pato_obo_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "curated_go_annotations = data.get_annotations_dictionary(\"GO\")\n",
    "curated_po_annotations = data.get_annotations_dictionary(\"PO\")\n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>freq</th>\n",
       "      <th>term_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>1481</td>\n",
       "      <td>protein binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>921</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0009507</td>\n",
       "      <td>601</td>\n",
       "      <td>chloroplast</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>GO:0005886</td>\n",
       "      <td>573</td>\n",
       "      <td>plasma membrane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>GO:0005829</td>\n",
       "      <td>476</td>\n",
       "      <td>cytosol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>323</td>\n",
       "      <td>molecular_function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>GO:0005737</td>\n",
       "      <td>320</td>\n",
       "      <td>cytoplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>GO:0009793</td>\n",
       "      <td>279</td>\n",
       "      <td>embryo development ending in seed dormancy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>266</td>\n",
       "      <td>mitochondrion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GO:0006355</td>\n",
       "      <td>256</td>\n",
       "      <td>regulation of transcription, DNA-templated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0009570</td>\n",
       "      <td>243</td>\n",
       "      <td>chloroplast stroma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>GO:0009506</td>\n",
       "      <td>232</td>\n",
       "      <td>plasmodesma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GO:0009941</td>\n",
       "      <td>214</td>\n",
       "      <td>chloroplast envelope</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>GO:0044212</td>\n",
       "      <td>212</td>\n",
       "      <td>transcription regulatory region DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>GO:0005783</td>\n",
       "      <td>194</td>\n",
       "      <td>endoplasmic reticulum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>GO:0009737</td>\n",
       "      <td>190</td>\n",
       "      <td>response to abscisic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GO:0009651</td>\n",
       "      <td>186</td>\n",
       "      <td>response to salt stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>GO:0005794</td>\n",
       "      <td>172</td>\n",
       "      <td>Golgi apparatus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>GO:0042742</td>\n",
       "      <td>160</td>\n",
       "      <td>defense response to bacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GO:0003700</td>\n",
       "      <td>155</td>\n",
       "      <td>DNA-binding transcription factor activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        term_id  freq                                    term_name\n",
       "22   GO:0005515  1481                              protein binding\n",
       "2    GO:0005634   921                                      nucleus\n",
       "7    GO:0009507   601                                  chloroplast\n",
       "44   GO:0005886   573                              plasma membrane\n",
       "64   GO:0005829   476                                      cytosol\n",
       "39   GO:0003674   323                           molecular_function\n",
       "69   GO:0005737   320                                    cytoplasm\n",
       "179  GO:0009793   279   embryo development ending in seed dormancy\n",
       "48   GO:0005739   266                                mitochondrion\n",
       "24   GO:0006355   256   regulation of transcription, DNA-templated\n",
       "8    GO:0009570   243                           chloroplast stroma\n",
       "89   GO:0009506   232                                  plasmodesma\n",
       "20   GO:0009941   214                         chloroplast envelope\n",
       "35   GO:0044212   212  transcription regulatory region DNA binding\n",
       "106  GO:0005783   194                        endoplasmic reticulum\n",
       "293  GO:0009737   190                    response to abscisic acid\n",
       "41   GO:0009651   186                      response to salt stress\n",
       "51   GO:0005794   172                              Golgi apparatus\n",
       "152  GO:0042742   160                defense response to bacterium\n",
       "26   GO:0003700   155    DNA-binding transcription factor activity"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Which GO terms are used to annotate the most genes in this dataset?\n",
    "term_id_to_ids = defaultdict(list)\n",
    "for i,term_id_list in curated_go_annotations.items():\n",
    "    for term_id in term_id_list:\n",
    "        term_id_to_ids[term_id].append(i)\n",
    "term_id_to_num_ids = {k:len(v) for k,v in term_id_to_ids.items()}\n",
    "terms_df = pd.DataFrame(term_id_to_num_ids.items(), columns=[\"term_id\", \"freq\"])\n",
    "\n",
    "def get_term_name(ont,i):\n",
    "    try:\n",
    "        return(ont[i].name)\n",
    "    except:\n",
    "        return(\"\")\n",
    "\n",
    "terms_df[\"term_name\"] = terms_df[\"term_id\"].map(lambda x: get_term_name(go,x))\n",
    "terms_df.sort_values(by=\"freq\", ascending=False, inplace=True)\n",
    "terms_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the group be ones that have that GO term annotation.\n",
    "#go_term_id_of_interest = \"GO:0009640\"\n",
    "#gene_ids_in_this_pathway = [k for k,v in curated_go_annotations.items() if go_term_id_of_interest in v]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_name</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>num_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gluconeogenesis I</td>\n",
       "      <td>GLUCONEO-PWY</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>suberin monomers biosynthesis</td>\n",
       "      <td>PWY-1121</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>phenylpropanoid biosynthesis</td>\n",
       "      <td>PWY-361</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sporopollenin precursors biosynthesis</td>\n",
       "      <td>PWY-6733</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>very long chain fatty acid biosynthesis I</td>\n",
       "      <td>PWY-5080</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>starch biosynthesis</td>\n",
       "      <td>PWY-622</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>indole-3-acetate biosynthesis II</td>\n",
       "      <td>PWY-581</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gluconeogenesis III</td>\n",
       "      <td>PWY66-399</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>phosphatidylcholine acyl editing</td>\n",
       "      <td>PWY-6803</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>flavonoid biosynthesis</td>\n",
       "      <td>PWY1F-FLAVSYN</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>sucrose degradation II (sucrose synthase)</td>\n",
       "      <td>PWY-3801</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>glycolysis I (from glucose 6-phosphate)</td>\n",
       "      <td>GLYCOLYSIS</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3-phosphoinositide biosynthesis</td>\n",
       "      <td>PWY-6352</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cutin biosynthesis</td>\n",
       "      <td>PWY-321</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>L-leucine biosynthesis</td>\n",
       "      <td>LEUSYN-PWY</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 pathway_name     pathway_id  num_genes\n",
       "0                           gluconeogenesis I   GLUCONEO-PWY         25\n",
       "1               suberin monomers biosynthesis       PWY-1121         24\n",
       "2                phenylpropanoid biosynthesis        PWY-361         24\n",
       "3       sporopollenin precursors biosynthesis       PWY-6733         23\n",
       "4   very long chain fatty acid biosynthesis I       PWY-5080         22\n",
       "5                         starch biosynthesis        PWY-622         20\n",
       "6            indole-3-acetate biosynthesis II        PWY-581         20\n",
       "7                         gluconeogenesis III      PWY66-399         20\n",
       "8            phosphatidylcholine acyl editing       PWY-6803         20\n",
       "9                      flavonoid biosynthesis  PWY1F-FLAVSYN         20\n",
       "10  sucrose degradation II (sucrose synthase)       PWY-3801         19\n",
       "11    glycolysis I (from glucose 6-phosphate)     GLYCOLYSIS         17\n",
       "12            3-phosphoinositide biosynthesis       PWY-6352         17\n",
       "13                         cutin biosynthesis        PWY-321         17\n",
       "14                     L-leucine biosynthesis     LEUSYN-PWY         16"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the mappings from this dataset to PlantCyc information.\n",
    "#pmn_pathways_filename = \"../data/pickles/groupings_from_pmn_pathways.pickle\"                        \n",
    "#groups = load_from_pickle(pmn_pathways_filename)\n",
    "#id_to_group_ids, group_id_to_ids = groups.get_groupings_for_dataset(data)\n",
    "\n",
    "\n",
    "# Reading in the dataset of groupings for pathways in PlantCyc.\n",
    "plantcyc_name_mapping = {row.group_id:row.group_name for row in pd.read_csv(plantcyc_pathways_names_path).itertuples()}\n",
    "plantcyc_grouping = Groupings(plantcyc_pathways_path, plantcyc_name_mapping)\n",
    "id_to_group_ids, group_id_to_ids = plantcyc_grouping.get_groupings_for_dataset(data)\n",
    "\n",
    "# Look at which pathways are best represented in this dataset.\n",
    "pathways_sorted = sorted(group_id_to_ids.items(), key=lambda item: len(item[1]), reverse=True)\n",
    "pathways_sorted_lengths = [(i,len(l)) for (i,l) in pathways_sorted]\n",
    "pathways_df = pd.DataFrame(pathways_sorted_lengths, columns=[\"pathway_id\",\"num_genes\"])\n",
    "pathways_df[\"pathway_name\"] = pathways_df[\"pathway_id\"].map(lambda x: plantcyc_grouping.get_long_name(x))\n",
    "pathways_df = pathways_df[[\"pathway_name\",\"pathway_id\",\"num_genes\"]]\n",
    "pathways_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1045, 1211, 1222, 1336, 1337, 1338, 1742, 2602, 5467, 5479]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# For some example pathway to use.\n",
    "#pathway_id = \"PWY-361\"\n",
    "pathway_id = \"PWY-581\"\n",
    "#pathway_id = \"PWY-1121\"\n",
    "pathway_id = \"PWY-695\"\n",
    "gene_ids_in_this_pathway = group_id_to_ids[pathway_id]\n",
    "gene_ids_in_this_pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "wordcloud = defaultdict(list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PO:0000258</td>\n",
       "      <td>root cortex</td>\n",
       "      <td>30</td>\n",
       "      <td>5156</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.001604</td>\n",
       "      <td>0.032071</td>\n",
       "      <td>True</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PO:0020031</td>\n",
       "      <td>radicle</td>\n",
       "      <td>21</td>\n",
       "      <td>5165</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000826</td>\n",
       "      <td>0.016511</td>\n",
       "      <td>True</td>\n",
       "      <td>0.342031</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PO:0006203</td>\n",
       "      <td>pericycle</td>\n",
       "      <td>31</td>\n",
       "      <td>5155</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.001705</td>\n",
       "      <td>0.034105</td>\n",
       "      <td>True</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PO:0000045</td>\n",
       "      <td>embryo root</td>\n",
       "      <td>27</td>\n",
       "      <td>5159</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.001317</td>\n",
       "      <td>0.026333</td>\n",
       "      <td>True</td>\n",
       "      <td>0.236308</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PO:0005708</td>\n",
       "      <td>cortex</td>\n",
       "      <td>36</td>\n",
       "      <td>5150</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002259</td>\n",
       "      <td>0.045176</td>\n",
       "      <td>True</td>\n",
       "      <td>0.201662</td>\n",
       "      <td>*</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      term_id   term_label  genes_with  genes_without  group_genes_with  group_genes_without   p_value  p_value_adj  significant  info_content significance\n",
       "2  PO:0000258  root cortex          30           5156                 2                    8  0.001604     0.032071         True      0.400000            *\n",
       "0  PO:0020031      radicle          21           5165                 2                    8  0.000826     0.016511         True      0.342031            *\n",
       "3  PO:0006203    pericycle          31           5155                 2                    8  0.001705     0.034105         True      0.300000            *\n",
       "1  PO:0000045  embryo root          27           5159                 2                    8  0.001317     0.026333         True      0.236308            *\n",
       "4  PO:0005708       cortex          36           5150                 2                    8  0.002259     0.045176         True      0.201662            *"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_po_annotations, gene_ids_in_this_pathway, po).head(20)\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: po.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# ns   P > 0.05\n",
    "# *    P  0.05\n",
    "# **   P  0.01\n",
    "# ***  P  0.001\n",
    "# **** P  0.0001\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in results.itertuples():\n",
    "    wordcloud[\"Weight\"].append(int(1/row.p_value_adj))\n",
    "    wordcloud[\"Word\"].append(\"{} ({})\".format(row.term_id,row.term_label))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>term_id</th>\n",
       "      <th>term_label</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>info_content</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0009688</td>\n",
       "      <td>abscisic acid biosynthetic process</td>\n",
       "      <td>8</td>\n",
       "      <td>5178</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1.022962e-11</td>\n",
       "      <td>2.045924e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0045549</td>\n",
       "      <td>9-cis-epoxycarotenoid dioxygenase activity</td>\n",
       "      <td>7</td>\n",
       "      <td>5179</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2.269638e-09</td>\n",
       "      <td>4.539277e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.545455</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0016106</td>\n",
       "      <td>sesquiterpenoid biosynthetic process</td>\n",
       "      <td>12</td>\n",
       "      <td>5174</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4.902689e-11</td>\n",
       "      <td>9.805378e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>0.440112</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0043289</td>\n",
       "      <td>apocarotenoid biosynthetic process</td>\n",
       "      <td>8</td>\n",
       "      <td>5178</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1.022962e-11</td>\n",
       "      <td>2.045924e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>0.408166</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0009687</td>\n",
       "      <td>abscisic acid metabolic process</td>\n",
       "      <td>17</td>\n",
       "      <td>5169</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.078039e-10</td>\n",
       "      <td>4.156078e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>0.408166</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0010436</td>\n",
       "      <td>carotenoid dioxygenase activity</td>\n",
       "      <td>7</td>\n",
       "      <td>5179</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2.269638e-09</td>\n",
       "      <td>4.539277e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.408166</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0006714</td>\n",
       "      <td>sesquiterpenoid metabolic process</td>\n",
       "      <td>23</td>\n",
       "      <td>5163</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>7.717983e-10</td>\n",
       "      <td>1.543597e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.391221</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0043288</td>\n",
       "      <td>apocarotenoid metabolic process</td>\n",
       "      <td>17</td>\n",
       "      <td>5169</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.078039e-10</td>\n",
       "      <td>4.156078e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>0.372397</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:1902645</td>\n",
       "      <td>tertiary alcohol biosynthetic process</td>\n",
       "      <td>8</td>\n",
       "      <td>5178</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1.022962e-11</td>\n",
       "      <td>2.045924e-10</td>\n",
       "      <td>True</td>\n",
       "      <td>0.357340</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GO:0016114</td>\n",
       "      <td>terpenoid biosynthetic process</td>\n",
       "      <td>48</td>\n",
       "      <td>5138</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.208569e-08</td>\n",
       "      <td>4.417137e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>0.281001</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:1902644</td>\n",
       "      <td>tertiary alcohol metabolic process</td>\n",
       "      <td>17</td>\n",
       "      <td>5169</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.078039e-10</td>\n",
       "      <td>4.156078e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>0.252327</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GO:0006721</td>\n",
       "      <td>terpenoid metabolic process</td>\n",
       "      <td>71</td>\n",
       "      <td>5115</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1.395625e-07</td>\n",
       "      <td>2.791250e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.238738</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GO:0008299</td>\n",
       "      <td>isoprenoid biosynthetic process</td>\n",
       "      <td>54</td>\n",
       "      <td>5132</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>3.834365e-08</td>\n",
       "      <td>7.668731e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>0.212424</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0046165</td>\n",
       "      <td>alcohol biosynthetic process</td>\n",
       "      <td>14</td>\n",
       "      <td>5172</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>9.197935e-11</td>\n",
       "      <td>1.839587e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>0.207109</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GO:0016702</td>\n",
       "      <td>oxidoreductase activity, acting on single dono...</td>\n",
       "      <td>15</td>\n",
       "      <td>5171</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2.646116e-08</td>\n",
       "      <td>5.292231e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>0.203094</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GO:1901617</td>\n",
       "      <td>organic hydroxy compound biosynthetic process</td>\n",
       "      <td>65</td>\n",
       "      <td>5121</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>9.187433e-08</td>\n",
       "      <td>1.837487e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.184867</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GO:0006720</td>\n",
       "      <td>isoprenoid metabolic process</td>\n",
       "      <td>77</td>\n",
       "      <td>5109</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2.051176e-07</td>\n",
       "      <td>4.102351e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.178260</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GO:0016701</td>\n",
       "      <td>oxidoreductase activity, acting on single dono...</td>\n",
       "      <td>16</td>\n",
       "      <td>5170</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>3.304581e-08</td>\n",
       "      <td>6.609162e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>0.145642</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GO:0051213</td>\n",
       "      <td>dioxygenase activity</td>\n",
       "      <td>28</td>\n",
       "      <td>5158</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2.425545e-07</td>\n",
       "      <td>4.851090e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>0.132669</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GO:0006066</td>\n",
       "      <td>alcohol metabolic process</td>\n",
       "      <td>34</td>\n",
       "      <td>5152</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4.481532e-09</td>\n",
       "      <td>8.963063e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>0.128422</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       term_id                                         term_label  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant  info_content significance\n",
       "2   GO:0009688                 abscisic acid biosynthetic process           8           5178                 5                    5  1.022962e-11  2.045924e-10         True      0.545455         ****\n",
       "9   GO:0045549         9-cis-epoxycarotenoid dioxygenase activity           7           5179                 4                    6  2.269638e-09  4.539277e-08         True      0.545455         ****\n",
       "3   GO:0016106               sesquiterpenoid biosynthetic process          12           5174                 5                    5  4.902689e-11  9.805378e-10         True      0.440112         ****\n",
       "0   GO:0043289                 apocarotenoid biosynthetic process           8           5178                 5                    5  1.022962e-11  2.045924e-10         True      0.408166         ****\n",
       "6   GO:0009687                    abscisic acid metabolic process          17           5169                 5                    5  2.078039e-10  4.156078e-09         True      0.408166         ****\n",
       "10  GO:0010436                    carotenoid dioxygenase activity           7           5179                 4                    6  2.269638e-09  4.539277e-08         True      0.408166         ****\n",
       "8   GO:0006714                  sesquiterpenoid metabolic process          23           5163                 5                    5  7.717983e-10  1.543597e-08         True      0.391221         ****\n",
       "5   GO:0043288                    apocarotenoid metabolic process          17           5169                 5                    5  2.078039e-10  4.156078e-09         True      0.372397         ****\n",
       "1   GO:1902645              tertiary alcohol biosynthetic process           8           5178                 5                    5  1.022962e-11  2.045924e-10         True      0.357340         ****\n",
       "12  GO:0016114                     terpenoid biosynthetic process          48           5138                 5                    5  2.208569e-08  4.417137e-07         True      0.281001         ****\n",
       "7   GO:1902644                 tertiary alcohol metabolic process          17           5169                 5                    5  2.078039e-10  4.156078e-09         True      0.252327         ****\n",
       "17  GO:0006721                        terpenoid metabolic process          71           5115                 5                    5  1.395625e-07  2.791250e-06         True      0.238738         ****\n",
       "15  GO:0008299                    isoprenoid biosynthetic process          54           5132                 5                    5  3.834365e-08  7.668731e-07         True      0.212424         ****\n",
       "4   GO:0046165                       alcohol biosynthetic process          14           5172                 5                    5  9.197935e-11  1.839587e-09         True      0.207109         ****\n",
       "13  GO:0016702  oxidoreductase activity, acting on single dono...          15           5171                 4                    6  2.646116e-08  5.292231e-07         True      0.203094         ****\n",
       "16  GO:1901617      organic hydroxy compound biosynthetic process          65           5121                 5                    5  9.187433e-08  1.837487e-06         True      0.184867         ****\n",
       "18  GO:0006720                       isoprenoid metabolic process          77           5109                 5                    5  2.051176e-07  4.102351e-06         True      0.178260         ****\n",
       "14  GO:0016701  oxidoreductase activity, acting on single dono...          16           5170                 4                    6  3.304581e-08  6.609162e-07         True      0.145642         ****\n",
       "19  GO:0051213                               dioxygenase activity          28           5158                 4                    6  2.425545e-07  4.851090e-06         True      0.132669         ****\n",
       "11  GO:0006066                          alcohol metabolic process          34           5152                 5                    5  4.481532e-09  8.963063e-08         True      0.128422         ****"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = term_enrichment(curated_go_annotations, gene_ids_in_this_pathway, go).head(20)\n",
    "\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "\n",
    "\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "results[\"info_content\"] = results[\"term_id\"].map(lambda x: go.ic(x))\n",
    "results.sort_values(by=\"info_content\", ascending=False, inplace=True)\n",
    "\n",
    "\n",
    "# This lambda won't work is passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>token</th>\n",
       "      <th>genes_with</th>\n",
       "      <th>genes_without</th>\n",
       "      <th>group_genes_with</th>\n",
       "      <th>group_genes_without</th>\n",
       "      <th>p_value</th>\n",
       "      <th>p_value_adj</th>\n",
       "      <th>significant</th>\n",
       "      <th>significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bps</td>\n",
       "      <td>14</td>\n",
       "      <td>5172</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2.090975e-08</td>\n",
       "      <td>4.181950e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ga</td>\n",
       "      <td>57</td>\n",
       "      <td>5129</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>4.944089e-08</td>\n",
       "      <td>9.888177e-07</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>inhibitor</td>\n",
       "      <td>60</td>\n",
       "      <td>5126</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>6.295554e-08</td>\n",
       "      <td>1.259111e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>paclobutrazol</td>\n",
       "      <td>27</td>\n",
       "      <td>5159</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2.124322e-07</td>\n",
       "      <td>4.248645e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>germination</td>\n",
       "      <td>407</td>\n",
       "      <td>4779</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>1.886132e-06</td>\n",
       "      <td>3.772265e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>far</td>\n",
       "      <td>56</td>\n",
       "      <td>5130</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>3.204647e-06</td>\n",
       "      <td>6.409293e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>aba</td>\n",
       "      <td>268</td>\n",
       "      <td>4918</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>3.576682e-06</td>\n",
       "      <td>7.153365e-05</td>\n",
       "      <td>True</td>\n",
       "      <td>****</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>synthesis</td>\n",
       "      <td>64</td>\n",
       "      <td>5122</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>5.312245e-06</td>\n",
       "      <td>1.062449e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>entirely</td>\n",
       "      <td>17</td>\n",
       "      <td>5169</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>5.754502e-06</td>\n",
       "      <td>1.150900e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>germinate</td>\n",
       "      <td>66</td>\n",
       "      <td>5120</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>5.969779e-06</td>\n",
       "      <td>1.193956e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>suppressed</td>\n",
       "      <td>69</td>\n",
       "      <td>5117</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>7.066822e-06</td>\n",
       "      <td>1.413364e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>after</td>\n",
       "      <td>324</td>\n",
       "      <td>4862</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>1.060001e-05</td>\n",
       "      <td>2.120002e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>were</td>\n",
       "      <td>356</td>\n",
       "      <td>4830</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>1.813623e-05</td>\n",
       "      <td>3.627246e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>lover</td>\n",
       "      <td>2</td>\n",
       "      <td>5184</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>1.996396e-05</td>\n",
       "      <td>3.992792e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>almost</td>\n",
       "      <td>103</td>\n",
       "      <td>5083</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>3.245818e-05</td>\n",
       "      <td>6.491635e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>dormancy</td>\n",
       "      <td>34</td>\n",
       "      <td>5152</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>3.855138e-05</td>\n",
       "      <td>7.710276e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>was</td>\n",
       "      <td>414</td>\n",
       "      <td>4772</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>4.271760e-05</td>\n",
       "      <td>8.543519e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>treatment</td>\n",
       "      <td>145</td>\n",
       "      <td>5041</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>1.192348e-04</td>\n",
       "      <td>2.384697e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>germinated</td>\n",
       "      <td>52</td>\n",
       "      <td>5134</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>1.278097e-04</td>\n",
       "      <td>2.556195e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>resistant</td>\n",
       "      <td>308</td>\n",
       "      <td>4878</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1.506122e-04</td>\n",
       "      <td>3.012244e-03</td>\n",
       "      <td>True</td>\n",
       "      <td>**</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            token  genes_with  genes_without  group_genes_with  group_genes_without       p_value   p_value_adj  significant significance\n",
       "0             bps          14           5172                 4                    6  2.090975e-08  4.181950e-07         True         ****\n",
       "1              ga          57           5129                 5                    5  4.944089e-08  9.888177e-07         True         ****\n",
       "2       inhibitor          60           5126                 5                    5  6.295554e-08  1.259111e-06         True         ****\n",
       "3   paclobutrazol          27           5159                 4                    6  2.124322e-07  4.248645e-06         True         ****\n",
       "4     germination         407           4779                 7                    3  1.886132e-06  3.772265e-05         True         ****\n",
       "5             far          56           5130                 4                    6  3.204647e-06  6.409293e-05         True         ****\n",
       "6             aba         268           4918                 6                    4  3.576682e-06  7.153365e-05         True         ****\n",
       "7       synthesis          64           5122                 4                    6  5.312245e-06  1.062449e-04         True          ***\n",
       "8        entirely          17           5169                 3                    7  5.754502e-06  1.150900e-04         True          ***\n",
       "9       germinate          66           5120                 4                    6  5.969779e-06  1.193956e-04         True          ***\n",
       "10     suppressed          69           5117                 4                    6  7.066822e-06  1.413364e-04         True          ***\n",
       "11          after         324           4862                 6                    4  1.060001e-05  2.120002e-04         True          ***\n",
       "12           were         356           4830                 6                    4  1.813623e-05  3.627246e-04         True          ***\n",
       "13          lover           2           5184                 2                    8  1.996396e-05  3.992792e-04         True          ***\n",
       "14         almost         103           5083                 4                    6  3.245818e-05  6.491635e-04         True          ***\n",
       "15       dormancy          34           5152                 3                    7  3.855138e-05  7.710276e-04         True          ***\n",
       "16            was         414           4772                 6                    4  4.271760e-05  8.543519e-04         True          ***\n",
       "17      treatment         145           5041                 4                    6  1.192348e-04  2.384697e-03         True           **\n",
       "18     germinated          52           5134                 3                    7  1.278097e-04  2.556195e-03         True           **\n",
       "19      resistant         308           4878                 5                    5  1.506122e-04  3.012244e-03         True           **"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = token_enrichment(texts, gene_ids_in_this_pathway).head(20)\n",
    "\n",
    "\n",
    "threshold = 0.05\n",
    "results[\"p_value_adj\"] = multipletests(results[\"p_value\"].values, method='bonferroni')[1]\n",
    "results[\"significant\"] = results[\"p_value_adj\"] < threshold\n",
    "results = results.loc[results[\"significant\"]==True]\n",
    "\n",
    "\n",
    "# This lambda won't work if passed a value greater than the minimum p-value for significance defined here.\n",
    "significance_levels = {0.05:\"*\", 0.01:\"**\", 0.001:\"***\", 0.0001:\"****\"}\n",
    "get_level = lambda x: significance_levels[min([level for level in significance_levels.keys() if x <= level])]\n",
    "results[\"significance\"] = results[\"p_value_adj\"].map(get_level)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in results.itertuples():\n",
    "    wordcloud[\"Weight\"].append(int(1/row.p_value_adj))\n",
    "    wordcloud[\"Word\"].append(row.token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Weight</th>\n",
       "      <th>Word</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31</td>\n",
       "      <td>PO:0000258 (root cortex)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>60</td>\n",
       "      <td>PO:0020031 (radicle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>29</td>\n",
       "      <td>PO:0006203 (pericycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>37</td>\n",
       "      <td>PO:0000045 (embryo root)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>PO:0005708 (cortex)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2391228</td>\n",
       "      <td>bps</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1011308</td>\n",
       "      <td>ga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>794211</td>\n",
       "      <td>inhibitor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>235369</td>\n",
       "      <td>paclobutrazol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>26509</td>\n",
       "      <td>germination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15602</td>\n",
       "      <td>far</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>13979</td>\n",
       "      <td>aba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>9412</td>\n",
       "      <td>synthesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>8688</td>\n",
       "      <td>entirely</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>8375</td>\n",
       "      <td>germinate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>7075</td>\n",
       "      <td>suppressed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>4716</td>\n",
       "      <td>after</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2756</td>\n",
       "      <td>were</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2504</td>\n",
       "      <td>lover</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1540</td>\n",
       "      <td>almost</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1296</td>\n",
       "      <td>dormancy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1170</td>\n",
       "      <td>was</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>419</td>\n",
       "      <td>treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>391</td>\n",
       "      <td>germinated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>331</td>\n",
       "      <td>resistant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Weight                      Word\n",
       "0        31  PO:0000258 (root cortex)\n",
       "1        60      PO:0020031 (radicle)\n",
       "2        29    PO:0006203 (pericycle)\n",
       "3        37  PO:0000045 (embryo root)\n",
       "4        22       PO:0005708 (cortex)\n",
       "5   2391228                       bps\n",
       "6   1011308                        ga\n",
       "7    794211                 inhibitor\n",
       "8    235369             paclobutrazol\n",
       "9     26509               germination\n",
       "10    15602                       far\n",
       "11    13979                       aba\n",
       "12     9412                 synthesis\n",
       "13     8688                  entirely\n",
       "14     8375                 germinate\n",
       "15     7075                suppressed\n",
       "16     4716                     after\n",
       "17     2756                      were\n",
       "18     2504                     lover\n",
       "19     1540                    almost\n",
       "20     1296                  dormancy\n",
       "21     1170                       was\n",
       "22      419                 treatment\n",
       "23      391                germinated\n",
       "24      331                 resistant"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(wordcloud).to_csv(os.path.join(OUTPUT_DIR, \"{}_word_cloud.csv\".format(pathway_id)), index=False)\n",
    "pd.DataFrame(wordcloud)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "corn",
   "language": "python",
   "name": "corn"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
